### PR TITLE
Torch callbacks

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR proposes to add callbacks to the `TorchModel`:

- general interface `BaseCallback`: a class that implements a `on_iter_end` method

- `LogCallback`, `EarlyStopping`, `ReduceLROnPlateau`, as well as `PlateauCallback`, are already implemented

Currently, most of the docstrings are unfinished. I need feedback on functionality (and parameters) to make them clear and concise. Also, I am really unsatisfied with the implementation of the `on_plateau` condition in both `Pytorch` and `Keras`, and I am yet to be satisfied with the implementation of mine. Suggestions are welcome!

### CODE REVIEW

File: batchflow/models/torch/callbacks/plateau.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    def on_iter_end(self, **kwargs):
+        """ Perform the callback action, if on plateau. """
+        _ = kwargs
+
+        if self.plateau():
+            self.action()
+            self.cooldown_counter = self.cooldown - 1
+            self.stream(f'{self.__class__.__name__} at iteration {self.model.iteration}')
+
+    def action(self):
```
Comment (anton-br) [inline batchflow/models/torch/callbacks/plateau.py:55]: tell in docs what is a purpose of the action, not about implementation. And raise NotImplementedError here

File: batchflow/models/torch/callbacks/plateau.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+""" Callbacks to perform on loss plateau. """
+import numpy as np
+
+from .base import BaseCallback
+
+
+
+class PlateauCallback(BaseCallback):
```
Comment (anton-br) [inline batchflow/models/torch/callbacks/plateau.py:8]: if this class dont have an action to do. what about rename it to BasePlateauCallback class?

File: batchflow/models/torch/callbacks/base.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    def set_model(self, model):
+        self.model = model
+        return self
+
+    @abstractmethod
+    def on_iter_end(self, **kwargs):
+        """ !!. """
+
+
+class LogCallback(BaseCallback):
```
Comment (anton-br) [inline batchflow/models/torch/callbacks/base.py:44]: what about the tutorial in order to show how to use the callbacks?

File: batchflow/models/torch/callbacks/base.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+            msg = f'{timestamp}   {i:5}: {avg_loss:6.6f}'
+
+            if self.verbosity >= 2:
+                # Monitor resources
+                memory = round(USSMonitor.get_usage(), 2)
+                gpu_memory = GPUMemoryMonitor.get_usage()
+                gpu_utilization = GPUMonitor.get_usage()
+
+                msg += f' | {memory:6.3f} : '
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/torch/callbacks/base.py:89]: We already have them, and it was discussed in some of the previous PRs

File: batchflow/models/torch/callbacks/base.py (hunk 3 above)
Comment (SergeyTsimfer) [reply batchflow/models/torch/callbacks/base.py:44]: I actually believe that the time has come for a better `Torch` tutorial overall: a more advanced model (certainly not VGG), rich dataset (certainly not MNIST), more features from model-related like callbacks, microbatch to data augmentations. For now, I suggest this PR does the callback feature, and the demo notebook comes later

File: batchflow/models/torch/callbacks/base.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            return lambda msg: file_print(stream, msg)
+        raise TypeError('`Stream` argument must be either None, callable or string.')
+
+    def set_model(self, model):
+        self.model = model
+        return self
+
+    @abstractmethod
+    def on_iter_end(self, **kwargs):
+        """ !!. """
```
[the remark this answers is not in the document]
Comment (roman-kh) [reply batchflow/models/torch/callbacks/base.py:41]: Agree

File: batchflow/models/torch/base.py [hunk 6]
```diff
@@ -904,6 +915,8 @@ def train(self, *args, feed_dict=None, fetches=None, use_lock=True, profile=Fals
             'actual_model_inputs_shape': [get_shape(item) for item in _inputs],
             'actual_model_outputs_shape': get_shape(_targets),
         })
+
+        [callback.on_iter_end() for callback in self.callbacks]
```
Comment (a-arefina) [inline batchflow/models/torch/base.py:919]: why list comprehension?

File: batchflow/models/torch/callbacks/base.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        if stream is None:
+            return lambda *args: args
+
+        if callable(stream):
+            return stream
+
+        if isinstance(stream, str):
+            with open(stream, 'w') as _:
+                pass
+            return lambda msg: file_print(stream, msg)
```
Comment (a-arefina) [inline batchflow/models/torch/callbacks/base.py:40]: file reopening might be inefficient

File: batchflow/models/torch/callbacks/base.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    microbatch : bool
+        Whether to log size and number of microbatches.
+    """
+    def __init__(self, stream=None, frequency=1, resources=True, shapes=False, microbatch=False):
+        self.frequency = frequency
+        self.resources = resources
+        self.shapes = shapes
+        self.microbatch = microbatch
+
+        self.header = self.make_header()
```
Comment (a-arefina) [inline batchflow/models/torch/callbacks/base.py:83]: do you really need `header` to be an attribute? will you need it after writing it to the log?

File: batchflow/models/torch/callbacks/plateau.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    Parameters
+    ----------
+    mode : str
+        Mode of computing whether the loss has plateaued or not.
+        If `complex`, then we compare means of loss over the last `patience` / 2 iterations and the previous last.
+        If `pytorch` or `keras`, we compare the best (minimum) value over the last `patience`
+        iterations with the best value overall.
+        If `mean`, we compare the mean value over the last `patience` iterations with the best value overall.
+    patience : int
+        Length of analyzed interval. The bigger, the more stabilized loss should be to considered to be on plateau.
```
Comment (a-arefina) [inline batchflow/models/torch/callbacks/plateau.py:99]: I'd rather give a link to the docstring of `PlateauCallback` to reduce editing when `mode` description changes

**Pull request merged**